### PR TITLE
Wrap address book creation causal read in a transaction

### PR DESCRIPTION
**Pull request by ChristophWurst**
Found in the wild on an instance that likely used a DB cluster with read/write splits. This means the written data is not immediately readable if the read node is updated asynchronously.

The proper fix would be to migrate the code over to a wrapper and not fetch data that was just written. Until then we can wrap the two operation in a transaction, which makes the database run both queries on the write ("master") node.

Ref https://github.com/nextcloud/groupware/issues/24

### CODE REVIEW

**Comment by tcitworld** [comment]
Doesn't that apply to every use of `getLastInsertId()` that isn't inside a transaction? I could 4 others just in the `dav` app that look pretty similar.

**Comment by ChristophWurst** [comment]
Absolutely. I should probably wrap them as well.

**Review by st3iny — APPROVED** [review]
:rocket:

**Comment by ChristophWurst** [comment]
> Doesn't that apply to every use of `getLastInsertId()` that isn't inside a transaction? I count 4 others just in the `dav` app that look pretty similar.

https://github.com/nextcloud/server/pull/34415

**Review by tcitworld — APPROVED** [review]

File: apps/dav/lib/CardDAV/CardDavBackend.php [hunk 1]
```diff
@@ -420,20 +424,26 @@ public function createAddressBook($principalUri, $url, array $properties) {
...
+			$query->insert('addressbooks')
+				->values([
+					'uri' => $query->createParameter('uri'),
+					'displayname' => $query->createParameter('displayname'),
+					'description' => $query->createParameter('description'),
+					'principaluri' => $query->createParameter('principaluri'),
+					'synctoken' => $query->createParameter('synctoken'),
+				])
+				->setParameters($values)
+				->execute();
```
[the remark this answers is not in the document]
Comment (ChristophWurst) [reply apps/dav/lib/CardDAV/CardDavBackend.php:438]: One fix at a time :P

**Review by miaulalala — APPROVED** [review]

**Pull request merged**